NAVAND-714 - Added Waypoints to NavigationRoute; - RouteOptionsUpdater refactored;

RingerJK opened the pull request:
### Description

<!--
---------- CHECKLIST ----------
1. Add related labels (`bug`, `feature`, `new API(s)`, `SEMVER-MAJOR`, `needs-backporting`, etc.).
1. Adda a changelog entry under `Unreleased` tag or a `skip changelog` label if not applicable.
1. Update progress status on the project board.
1. Request a review from the team, if not a draft.
1. Add targeted milestone, when applicable.
1. Create ticket tracking addition of public documentation pages entry, when applicable.
-->

VysotskiVadim commented on libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/Waypoint.kt:
```diff
@@ -0,0 +1,27 @@
+    val name: String,
+    val target: Point?,
+) {
+    companion object {
+        const val REGULAR = 1
+        const val SILENT = 2
+        const val EV_CHARGING = 3
+    }
+
+    @Target(
+        AnnotationTarget.PROPERTY,
+        AnnotationTarget.VALUE_PARAMETER,
+        AnnotationTarget.FUNCTION,
+        AnnotationTarget.TYPE
+    )
+    @Retention(AnnotationRetention.BINARY)
+    @IntDef(REGULAR, SILENT, EV_CHARGING)
+    annotation class Type
+}
```
Why not enum? It's an internal API, we can break it, don't we?

RingerJK replied:
it might become public, so I make it forward compatible

dzinad commented:
Which ticket is it?

RingerJK commented:
> Which ticket is it?

Added ticket in the title

dzinad commented:
BillingController uses `RouteProgress#remainingWaypoints`. Should its logic stay the same? I'd think yes, but want to clarify.

RingerJK commented on libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/extensions/WaypointEx.kt:
```diff
@@ -0,0 +1,43 @@
+ * Return waypoints that are requested explicitly
+ */
+fun List<Waypoint>.requestedWaypoints(): List<Waypoint> =
+    this.filter { it.type != Waypoint.EV_CHARGING }
```
yes, I missed it, good catch

dzinad replied:
That you only accept some types, not filter them out.
`When` addresses it and even makes it more explicit.

LukasPaczos commented on libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/utils/RouterEx.kt:
```diff
@@ -15,3 +17,5 @@ fun com.mapbox.navigation.base.route.RouterOrigin.mapToNativeRouteOrigin(): Rout
         is com.mapbox.navigation.base.route.RouterOrigin.Custom -> RouterOrigin.CUSTOM
     }
+
+fun NavigationRoute._waypoints(): List<Waypoint> = waypoints
```
```suggestion
fun NavigationRoute.waypointsList(): List<Waypoint> = waypoints
```
nit

RingerJK replied:
added `_` to mark it kinda `internal`. So may be
```suggestion
fun NavigationRoute.internalWaypoints(): List<Waypoint> = waypoints
```

LukasPaczos commented on libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt:
```diff
@@ -203,6 +208,38 @@ class NavigationRoute internal constructor(
             }
         }
 
+        internal fun create(
```
Why are we adding the function in this PR if it's only used in tests and actually crashes if tried to use in main codebase? Let's re-add it when we can actually take advantage of the native logic.

RingerJK replied:
https://github.com/mapbox/mapbox-navigation-android/pull/6516

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:
```diff
@@ -379,7 +380,7 @@ internal class MapboxTripSession(
 
         private fun calculateRemainingWaypoints(tripStatus: TripStatus): Int {
-            val routeCoordinates = tripStatus.route?.routeOptions?.coordinatesList()
+            val routeCoordinates = tripStatus.route?._waypoints()
```
This would now also report silent waypoints which is a breaking change. It should only report regular and EV waypoints, so actually the ones that create legs.

RingerJK replied:
why it's breaking changes? We haven't had EV waypoints, as they appeared, now is a part of the API if anybody uses it

LukasPaczos replied:
I'm talking about silent waypoints. They are not really waypoints and should not be part of the remaining waypoints count.

RingerJK replied:
good catch 👍

RingerJK replied:
Also, silent and Ev waypoints are under `NavigationStatus#getNextWaypointIndex`(which is root of the `remainingWaypoints` value)

LukasPaczos replied:
I think it's rather confusing that we call this a `remainingWaypoint` when a silent waypoint is not truly a waypoint in the Directions API terminology.

I'm worried that if someone used to use `remainingWaypoints` to compare it with `RouteOptions#coordinates` (as that's the only practical usage because `remainingWaypoints` contains silent waypoints) then the proposed change will break that logic.

RingerJK replied:
from the SDK side, it seems like not a breaking change, it might be a misunderstanding between the SDK docs and customers. Let's considered it in the API doc and add the warning to the changelog. WDYT?

LukasPaczos replied:
Yea, this is fine, refs https://github.com/mapbox/mapbox-navigation-android/pull/6005#issuecomment-1292301010. But I agree that we should add a note in the docs that `remainingWaypoints` now might include more than requested waypoints if EV is used.

RingerJK commented on libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/extensions/WaypointEx.kt:
```diff
@@ -0,0 +1,43 @@
+ *
+ * For instance, EV waypoints are not requested explicitly, so they are not taken into account.
+ */
+fun indexOfNextRequestedCoordinate(
```
I get the name from `RouteOptions#coordinates` which is means the same here

dudeuter replied:
I understand `coordinates` as locations that a user is asking directions to, and `waypoints` as the actual points that have been map matched onto the road.

A "requested" waypoint would mean one that hasn't been discreetly added by the API.

RingerJK replied:
yes, that's actually what the fun is returned: then next `REQUESTSTED`coordinate  (from the list `RouteOptions#coorinates`). Added the note 
```md
See [RouteOptions.coordinatesList]
```

LukasPaczos commented on libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/extensions/WaypointEx.kt:
```diff
@@ -0,0 +1,43 @@
+@file:JvmName("WaypointEx")
```
mockk has an ability to mock extensions/static functions by function reference, we shouldn't need the JVM name

RingerJK replied:
actually we have to mock static cunction via `mockkStatic("package.name.fileName")`. For instance, have a look at `RouteOptionsUpdaterTest` 
![image](https://user-images.githubusercontent.com/8738058/198243099-c7d998d9-2357-4b5d-9cda-6c3a99aba43f.png). 

You can also find other samples in the SDK unit tests

LukasPaczos replied:
@RingerJK you can do:
```kotlin
mockkStatic(::indexOfNextRequestedCoordinate) {
    // ...
}
```
No need for `@file:JvmName("WaypointEx")`.

RingerJK replied:
looks nice, substitute with that 👍

RingerJK commented:
> 1. Docs should be updated to specify which waypoints these are;

if we are speaking about `Waypoint.kt` - I'll do it 👍 

> 2. An idea: would it make sense to add another field for the other type of waypoints? My concern is that we might have to apply the logic that's in RouteOptionsUpdater somewhere else if we use `RouteProgress#remainingWaypoints` somewhere else. Also the end user can use this property. Now that the waypoints are of different types, it would make sense to allow the user to choose which ones they'd like to use. TBH I don't really understand what's happening in `RouteOptionsUpdater`. What's refactoring and what's logic change. IMO it would be nice to separate: the logic change would be in `calculateRemainingWaypoints` or somewhere near it. WDYT?

The Waypoint class is internal itself so we're good to experiment and change it. 
- What fields would you have in mind to add for `other` types of waypoints? 
- how do you see customers might choose the particular type of waypoints? For now, they are mostly like passive observer  
- Yeah, I agree, the logic is a bit hard to get in `RouteOptionsUpdater`. Would it help if I move everything into separate functions? 

>BillingController uses RouteProgress#remainingWaypoints. Should its logic stay the same? I'd think yes, but want to clarify.

great catch actually. I would say it should be switched to the `Waypoint` as well (I see there's using `DirecionsResponce#DirectionsWaypoint`. @LukasPaczos do you have objections?

RingerJK commented on libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/Waypoint.kt:
```diff
@@ -0,0 +1,27 @@
+import com.mapbox.geojson.Point
+
+data class Waypoint(
+    val location: Point,
```
do you mean fakeing `DirectionsWaypoint` based on `Waypoint` instance?

RingerJK replied:
We don't always have `DirectionsWaypoint`: for instance, we don't have it when creating NavigationRoute based on DirecionsRoute(s) and a `Waypoint` object is created based on the NN's Waypoint.

LukasPaczos replied:
Okay, let's revisit later if needed.

LukasPaczos commented on libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/extensions/WaypointEx.kt:
```diff
@@ -0,0 +1,41 @@
+        .firstOrNull()
+        ?: return null
+
+    return waypoints.indexOf(nextRequestedWaypoint)
```
Please correct me if I'm wrong but it looks like the index returned here is the index in the `Waypoint` list which contains EV waypoints but we use this index to mutate the original request which does not cotain for EV waypoints. I think we're looking for something like this instead:
```suggestion
    return waypoints.requestedWaypoints().indexOf(nextRequestedWaypoint)
```

dzinad replied:
I think this method is about finding the right index in the passed list.

RingerJK replied:
you're both right 😄  I tried to avoid calling the second `requestedWaypoints` but missed that it's required to find the right index (tests become 🔴, so we're in safety 🥲  ). Already in the branch

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routeoptions/RouteOptionsUpdater.kt:
```diff
@@ -46,94 +47,110 @@ class RouteOptionsUpdater {
+                        }
+                    }
+                )
+                .approachesList(
```
Approaches should be specified per waypoint, not per coordinate (capturing from https://docs.mapbox.com/api/navigation/directions/#optional-parameters), so it should use `getUpdatedWaypointsList`.

LukasPaczos replied:
We can address separately, let's cut a ticket.

RingerJK replied:
`NAVAND-834`

LukasPaczos commented:
> great catch actually. I would say it should be switched to the Waypoint as well (I see there's using DirecionsResponce#DirectionsWaypoint. @LukasPaczos do you have objections?
>> We can switch the billing controller to use Waypoint but it should ignore silent waypoints, it's only interested in leg waypoints.

Sorry, this always surprises me time and time again that `DirectionsResponse#waypoints` contains silent waypoints. Since `DirectionsResponse#waypoints` contains all of regular, silent, and EV waypoints and `RouteProgress#remainingWaypoints` is already updated to account for EV waypoints, then we shouldn't need to make any changes. Unless I'm missing something @RingerJK?

dzinad commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routeoptions/RouteOptionsUpdater.kt:
```diff
@@ -46,94 +47,110 @@ class RouteOptionsUpdater {
-                    .waypointIndicesList(
-                        getUpdatedWaypointIndicesList(
-                            routeOptions.waypointIndicesList(),
-                            coordinatesList.size - remainingWaypoints - 1
```
Does this really correspond to nextCoordinateIndex?

dzinad commented:
> if we are speaking about Waypoint.kt - I'll do it 👍

No, I was talking about `RouteProgress#remainingWaypoints` doc.

> The Waypoint class is internal itself so we're good to experiment and change it.

Again, I was talking about `RouteProgress`.

> What fields would you have in mind to add for other types of waypoints?
how do you see customers might choose the particular type of waypoints? For now, they are mostly like passive observer

Requested waypoints and all waypoints? I don't know how they are used now. As far as I understood from the ticket description, the problem is that now in the response we may have more waypoints than in the request. Who knows hot this data will be used by users, why not expose both numbers?

> Yeah, I agree, the logic is a bit hard to get in RouteOptionsUpdater. Would it help if I move everything into separate functions?

Nah, not really. Let's leave as is.

RingerJK commented on libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/extensions/WaypointEx.kt:
```diff
@@ -0,0 +1,43 @@
+        ?: return null
+
+    return waypoints
+        .requestedWaypoints()
```
logistic I think might have a lot of waypoints (probably put to 100, but I don't have any statistics). 
Do you see how can we get the index of the next requested waypoint another way?

dzinad replied:
```
    var requestedIndex = 0
    waypoints.forEachIndexed { index, waypoint ->
        if (waypoint.type == Waypoint.REGULAR || waypoint.type == Waypoint.SILENT) {
            if (index >= waypoints.size - remainingWaypoints) {
                return requestedIndex
            }
            requestedIndex++
        }
    }
    return null
```

I think it would be equivalent.

RingerJK commented:
> Sorry, this always surprises me time and time again that `DirectionsResponse#waypoints` contains silent waypoints. Since `DirectionsResponse#waypoints` contains all of regular, silent, and EV waypoints and `RouteProgress#remainingWaypoints` is already updated to account for EV waypoints, then we shouldn't need to make any changes in the billing controller. Unless I'm missing something @RingerJK?

I would say that for now, we can keep it as is, and if the `Waypoint` _experiment_ becomes a success we can switch to that as a single source of truth. 

we also need utils methods for us and customers to work with waypoints, as it's always a challenge to handle them right (for instance getNextRequestedCoordinate might be hard to make on the application side)

@LukasPaczos I think it's a typo in the doc, shouldn't be silent, right?
https://github.com/mapbox/mapbox-navigation-android/blob/934e668205bf7688734db9ff00b199edd5055044/libnavigation-core/src/main/java/com/mapbox/navigation/core/accounts/BillingController.kt#L395-L403

RingerJK commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routeoptions/RouteOptionsUpdater.kt:
```diff
@@ -46,94 +47,110 @@ class RouteOptionsUpdater {
+                        mutableListOf<Double?>().also {
+                            it.addAll(
+                                radiusesList.subList(
+                                    nextCoordinateIndex - 1,
```
`nextCoordinateIndex` is receiving from `indexOfNextRequestedCoordinate`, which is always return value between [1; size_of_waypoint_list) (null is also possible, but it filtered out)

RingerJK commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routeoptions/RouteOptionsUpdater.kt:
```diff
@@ -46,94 +47,110 @@ class RouteOptionsUpdater {
-                                routeOptions.snappingIncludeClosuresList()
-                                    .withFirstTrue(remainingWaypoints, index, coordinatesList.size)
+                        mutableListOf<Double?>().also {
+                            it.addAll(
```
replaced by 
```kotlin
.radiusesList(
    let radiusesList@{
        val radiusesList = routeOptions.radiusesList()
        if (radiusesList.isNullOrEmpty()) {
            return@radiusesList emptyList<Double?>()
        }
        radiusesList.subList(
            nextCoordinateIndex - 1,
            coordinatesList.size
        )
    }
)
```

LukasPaczos commented:
> I think it's a typo in the doc, shouldn't be silent, right?

Yes :+1:

RingerJK commented:
Updates: 
- cut `NAVAND-825` to address https://github.com/mapbox/mapbox-navigation-android/pull/6005#issuecomment-1292548529. Let's discuss if we want to expose it in the ticket
- fixed docs in RouteProgress and BillingController;
- added `skip changelog`, as it fails for my changes.

RingerJK commented on libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/extensions/WaypointEx.kt:
```diff
@@ -0,0 +1,58 @@
+            Waypoint.REGULAR,
+            Waypoint.SILENT -> true
+            Waypoint.EV_CHARGING -> false
+            else -> false
```
the test does not guarantee that every place where it is used be updated along with new waypoint types, also it limits adding new fields to the companion object. 

let me know what you think: I added internal enum and keep `Def` to make it forward compatible (also, it's used by other modules) https://github.com/mapbox/mapbox-navigation-android/pull/6005/commits/0bd7ccf365b1b9ec8a09e4aabeed5aacd12efee3

dzinad replied:
It's not guaranteed that we'll use internalType, not type everywhere in our code. It's kind of a compromise though.

> the test does not guarantee that every place where it is used be updated along with new waypoint types, also it limits adding new fields to the companion object.

It was out of context perhaps, I suggested adding the parameterized test for `getRequestedWaypoints` and so on and check the input values. So that when you add a new value, the test will fail because you haven't tested this value.You'll look at the test and go on and fix the original method.

> also it limits adding new fields to the companion object

You can always add `+1` with a comment that it's for this field. I don't think there would be too many false positives especially considering the fact that it's easily fixed.

I'll leave the choice to you, both options would work fine for me.

RingerJK replied:
discussed: approaches are similar, keep using enums

RingerJK commented on libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/extensions/WaypointEx.kt:
```diff
@@ -0,0 +1,58 @@
+    val nextWaypointIndex = waypoints.size - remainingWaypoints
+    var requestedIndex = 0
+    waypoints.forEachIndexed { index, waypoint ->
+        if (waypoint.type == Waypoint.REGULAR || waypoint.type == Waypoint.SILENT) {
```
the same tried at https://github.com/mapbox/mapbox-navigation-android/pull/6005#discussion_r1007890349 😄

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:
```diff
@@ -379,7 +381,9 @@ internal class MapboxTripSession(
 
         private fun calculateRemainingWaypoints(tripStatus: TripStatus): Int {
-            val routeCoordinates = tripStatus.route?.routeOptions?.coordinatesList()
+            val routeCoordinates = tripStatus.route
+                ?.internalWaypoints()
+                ?.filter { it.isLegWaypoint() }
```
This now doesn't include silent waypoints but it did in the past, doesn't this break the `waypointsCount - nextWaypointIndex` calculation below since NN does include silent waypoints in `tripStatus.navigationStatus.nextWaypointIndex`?

RingerJK replied:
it's actually breaking the logic, I'll fix it and add tests 👍

LukasPaczos reviewed (commented):
LGTM but it'd be great to add an instrumentation test that verifies a re-route with EV routes.

LukasPaczos commented on CHANGELOG.md:
```diff
@@ -9,6 +9,8 @@ Mapbox welcomes participation and contributions from everyone.
 - Fixed an issue with `NavigationView` that caused road label position to not update in some cases. [#6508](https://github.com/mapbox/mapbox-navigation-android/pull/6508)
 - Fixed an issue where the SDK was trying to send telemetry events when telemetry is switched off globally. [#6512](https://github.com/mapbox/mapbox-navigation-android/pull/6512)
 
+:warning: `RouteProgress#remainingWaypoints` defines the amount of all remaining waypoints, including explicit (requested with `RouteOptions#coordinatesList`) and implicit (added on the fly, like Electric Vehicle waypoints - [see](https://docs.mapbox.com/api/navigation/directions/#electric-vehicle-routing)) ones.
```
If we're adding the test above then it'd be good to address this as well.

dzinad commented:
> LGTM but it'd be great to add an instrumentation test that verifies a re-route with EV routes.

Oh, I was going to add the test in https://mapbox.atlassian.net/browse/NAVAND-775. It's about alternatives but there is no basic test. It would be real nice if it appeared. :)

LukasPaczos commented on CHANGELOG.md:
```diff
@@ -24,6 +25,8 @@ This release depends on, and has been tested with, the following Mapbox dependen
 - Mapbox Java `v6.9.0-beta.1` ([release notes](https://github.com/mapbox/mapbox-java/releases/tag/v6.9.0-beta.1))
 - Mapbox Android Core `v5.0.2` ([release notes](https://github.com/mapbox/mapbox-events-android/releases/tag/core-5.0.2))
 
+:warning: `RouteProgress#remainingWaypoints` defines the amount of all remaining waypoints, include explicit (requested with `RouteOptions#coorinatesList`) and implicit (added on the fly, like Electric Vehicle waypoints - [see](https://docs.mapbox.com/api/navigation/directions/#electric-vehicle-routing)).
+:warning: `RouteProgress#remainingWaypoints` defines the amount of all remaining waypoints, including explicit (requested with `RouteOptions#coordinatesList`) and implicit (added on the fly, like Electric Vehicle waypoints - [see](https://docs.mapbox.com/api/navigation/directions/#electric-vehicle-routing)) ones.
```
Rebase issue.

LukasPaczos reviewed (approved):
We should be in a good place with the implementation (and unit tests) and this lands only in the alpha but let's prioritize the instrumentation tests right after this.

RingerJK commented:
> We should be in a good place with the implementation (and unit tests) and this lands only in the alpha but let's prioritize the instrumentation tests right after this.

`NAVAND-835`
